build.proj

```diff
@@ -62,7 +63,18 @@
     <MSBuild Projects="@(PkgProject)"
              Targets="Pack" />
   </Target>
-
+  
+  <ItemGroup>
+     <TestFile Include="d" Url="https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv"/>
+  </ItemGroup>
+  
+  <Target Name="DownloadExternalTestFiles" Condition="'$(RunTests)'=='true'">
```

Comment:
I think we should always do this, not just when `RunTests` is on.  That way I can still run tests from VS Test Explorer even if I just "build.cmd" from command line.

However, we should also have "Inputs and Outputs" hooked up, so that way the target is skipped if the files already exist.  check out [https://msdn.microsoft.com/en-us/library/ms171483.aspx](https://msdn.microsoft.com/en-us/library/ms171483.aspx) for how to make incremental builds work.  Let me know if you need help.

Reply:
Any chance you can help me with that?

``` <ItemGroup>
     <TestFile Include="winequality-white.csv" Url="https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv">
         <Result>$(MSBuildThisFileDirectory)test\data\external\winequality-white.csv</Result>
     </TestFile>
      <TestFile Include="winequality-red.csv" Url="https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv">
         <Result>$(MSBuildThisFileDirectory)test\data\external\winequality-red.csv</Result>
     </TestFile>
  </ItemGroup>
  
  <Target Name="DownloadExternalTestFiles" Inputs="@(TestFile)" Outputs="%(TestFile.Result)">
    <Message Importance="High" Text="Downloading external test files... %(TestFile.Result)" />
      <DownloadFilesFromUrl Items="@(TestFile)"
                           DestinationDir="test/data/external"
                           TreatErrorsAsWarnings="true"/>
  </Target>
```  
It keeps download files and i'm not sure how to debug it properly

Reply:
@eerhardt  Can you help me?

Reply:
The way Inputs and Outputs work is based on file timestamps.  You need an actual Input file, so MSBuild can compare its last write time to the output file's last write time.

Here when you specify:

`Inputs="@(TestFile)"`

It uses the `TestFile`'s ItemSpec (what is put into the `Include` attribute).  I'm assuming we don't have files named "winequality-white.csv", so MSBuild probably decides to always do the task, since it can't find the input file.

Another option is to check if the local file already exists, and if it does, then don't do the download.
Here's that approach we do in a different repo that is doing a similar thing (downloading files from the internet during the build, and caching them so they aren't downloaded for every build invocation):

https://github.com/dotnet/cli/blob/69062dda4edd7670cffa5c23876b38eccc130e09/build/Prepare.targets#L36-L56

The above will download any URL in the `@(_DownloadAndExtractItem)` Item.  And then where those items are defined:

https://github.com/dotnet/cli/blob/69062dda4edd7670cffa5c23876b38eccc130e09/build/BundledRuntimes.props#L61-L66

The item is "Condition"d on whether the file exists or not:

```
    <_DownloadAndExtractItem Include="CombinedSharedHostAndFrameworkArchive"
                             Condition="!Exists('$(CombinedSharedHostAndFrameworkArchive)')">
      <Url>$(CoreSetupRootUrl)$(MicrosoftNETCoreAppPackageVersion)/$(CombinedFrameworkHostCompressedFileName)$(CoreSetupBlobAccessTokenParam)</Url>
      <DownloadFileName>$(CombinedSharedHostAndFrameworkArchive)</DownloadFileName>
      <ExtractDestination>$(SharedFrameworkPublishDirectory)</ExtractDestination>
    </_DownloadAndExtractItem>
```

So if the `CombinedSharedHostAndFrameworkArchive` file already exists, then this item isn't included.  Thus, when all the files already exist locally, nothing is downloaded.